### PR TITLE
Fix BuiltMap equality for different keys with null values

### PR DESCRIPTION
**Pull request by hoc081098**
### 1. Fixes `BuiltMap` equality when two maps have:

- the same length;
- the same hash code;
- different keys; and
- `null` values.

Previously, equality only compared values using `operator []`. Since a missing
key and a key mapped to `null` both return `null`, maps with different keys
could incorrectly compare equal after passing the length and hash-code checks.

The regression test compares `{1: null}` and `{2: null}` using overridden,
matching hash codes. This ensures equality reaches the entry comparison and
verifies that the maps are not equal.

### 2. Validation:

- `dart format --output=none --set-exit-if-changed .`
- `dart analyze --fatal-infos`
- `dart test`

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/blob/main/docs/External-Package-Maintenance.md#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.

**Note**: The Dart team is trialing Gemini Code Assist. Don't take its comments as final Dart team feedback. Use the suggestions if they're helpful; otherwise, wait for a human reviewer.

</details>

### CODE REVIEW

File: built_types/built_collection/CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,7 @@
 - Fix unnecessary non-null assertions in internal copy-on-write collections.
 - Migrate to `package:dart_flutter_team_lints`.
 - Fix analysis issues and remove the lint suppressions that hid them.
+- Fix `BuiltMap` equality when maps with nullable values have different keys.
```
Comment (davidmorgan) [inline built_types/built_collection/CHANGELOG.md]: It's narrower than that, let's be specific, e.g.:

    - Fix `BuiltMap` equality incorrectly returning true for maps with null values,                                                                                                                                
      different keys and colliding hash codes.

File: built_types/built_collection/lib/src/map/built_map.dart [hunk 2]
```diff
@@ -104,6 +104,8 @@ abstract class BuiltMap<K, V> {
     if (other.hashCode != hashCode) return false;
     for (final key in keys) {
       if (other[key] != this[key]) return false;
+      // A lookup can't distinguish a missing key from one mapped to null.
+      if (!other._map.containsKey(key)) return false;
```
Comment (davidmorgan) [inline built_types/built_collection/lib/src/map/built_map.dart]: We can save a lookup for all cases but the null one:

```
    for (final key in keys) {                                                                                                                                                                                      
      final value = this[key];                                                                                                                                                                                     
      if (other[key] != value) return false;                                                                                                                                                                       
      // A lookup cannot distinguish a missing key from one mapped to null.                                                                                                                                        
      if (value == null && !other._map.containsKey(key)) return false;                                                                                                                                             
    }  
```

File: built_types/built_collection/test/map/built_map_test.dart [hunk 3]
```diff
@@ -267,10 +266,26 @@ void main() {
       );
     });
 
+    test('compares not equal to different keys with null values', () {
```
Comment (davidmorgan) [inline built_types/built_collection/test/map/built_map_test.dart:269]: Please also test two equal content maps with null values compare as equal.

**Review by davidmorgan — COMMENTED** [review]
This is a great find, thank you--just a few suggestions.

**Comment by hoc081098** [comment]
> This is a great find, thank you--just a few suggestions.

Thanks for the review!
I've addressed your suggestions in https://github.com/dart-lang/build/pull/5154/changes/440fa42738d8f8320ad1c8d8b72c537b3cfc4c08
